### DISCUSSION TITLE
New feature: window edge-snapping overlay

### DISCUSSION THREAD

**Discussion by tokyo4j** [labwc/labwc#1613] [Show and tell]
I'm working on showing overlay when the window is being snapped on output edges ([source](https://github.com/tokyo4j/labwc/tree/snap-to-edge-overlay)).

It works pretty well, but it has a fundamental problem with multi-monitor setup, where flashy overlay is shown every time the window is moved from an output to another.

https://github.com/labwc/labwc/assets/22029524/8529b381-ec9b-4d89-9692-b7c67a72c976

I came up with setting a timer and putting some delay before showing the overlay, but I'm afraid that it complicates the code a lot because we then have to consider edge-cases like view destruction before the timeout.

Do you have any ideas to fix/improve this problem?
Keep in mind that I'm not gonna insist on implementing this feature.

**Comment by ahesford** [comment]
This looks like a great addition.

It is exceedingly difficult to snap to a window edge when there is an adjacent output; you have to get the cursor pixel-perfect. It can also produce unexpected results, like trying to hit the right edge of the left display but instead snapping to the left edge of the right display.

With this in mind, I would just skip drawing the overlay when snapping to an edge that has an output next to it. (I might go further and disable cursor-based snap-zones for edges shared by multiple outputs, but that's a different discussion.) There's no need for exceedingly complex logic to capture what I assume is a rare case.

**Reply by Consolatis** [reply]
The pixel perfect positioning requirement is likely due to the default setting of `1` for `<snap><range>`. I'd imagine with something like `10` or `20` its also less of a flashing but more of an indicator. Using a timer might indeed be an option. AFAIR we already check on view destruction if that view is currently moved around and then call `interactive_cancel` and that one in turn also hides a potential overlay. So when hiding the overlay we could also just cancel the timer.

**Comment by tokyo4j** [comment]
Thank you for suggestions.

I agree with @Consolatis.
Snapping on edges with an adjacent output is a viable workflow for me. In addition to setting `<snapping><range>` to 10, I put waybar on the top edge with vertically-aligned multi-monitor setup as shown in the video below, so I can grab the window and release it on the panel to maximize the window easily.

I made a prototype that delays the overlay by 150ms using a timer. It turned out to be additional 45 lines of code and much cleaner than I initially expected. I was worried that the latency would degrade user experience, but 150ms seems to make a good balance between latency and blinking prevention. So I will implement with this approach.

Anyway, thank you for your help!

https://github.com/labwc/labwc/assets/22029524/2a69ed0b-8f22-4910-a65c-84881ba44ce9

**Comment by ahesford** [comment]
Looks good. I didn't even realize there was a `<snapping><range>` option!

**Comment by Consolatis** [comment]
@tokyo4j do you consider opening a PR with your changes?
Matching the region overlays with snapping overlays is something that was on the back of my mind for a long time.

A general idea: maybe we could create a `src/overlay.c` file which contains the overlay specific functionality and keeps `interactive.c` a bit more targeted at the actual movement actions.
